src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,390 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Libero\ContentApiBundle\Adapter;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\FetchMode;
+use Doctrine\DBAL\LockMode;
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\DBAL\Types\Type;
+use IteratorAggregate;
+use Libero\ContentApiBundle\Exception\ItemNotFound;
+use Libero\ContentApiBundle\Exception\UnexpectedVersionNumber;
+use Libero\ContentApiBundle\Exception\VersionNotFound;
+use Libero\ContentApiBundle\Model\ItemId;
+use Libero\ContentApiBundle\Model\ItemListPage;
+use Libero\ContentApiBundle\Model\Items;
+use Libero\ContentApiBundle\Model\ItemVersion;
+use Libero\ContentApiBundle\Model\ItemVersionNumber;
+use Traversable;
+use function array_column;
+use function array_map;
+use function array_pop;
+use function count;
+use function rewind;
+use function sprintf;
+
+final class DoctrineItems implements IteratorAggregate, Items
+{
+    private const TRAVERSING_LIMIT = 100;
+
+    private $connection;
+    private $tableItems;
+    private $tableVersions;
+
+    public function __construct(Connection $connection, string $tablePrefix)
+    {
+        $this->connection = $connection;
+        $this->tableItems = "{$tablePrefix}_items";
+        $this->tableVersions = "{$tablePrefix}_versions";
+    }
+
+    public function add(ItemVersion $item) : void
+    {
+        $this->connection->transactional(
+            function () use ($item) : void {
+                try {
+                    $current = $this->getCurrentVersion($item->getId(), LockMode::PESSIMISTIC_READ);
```

Comment:
I have to do some experiments with one of the RDBMS underneath, but it seems that both use cases of add/remove modify `item.current_version` so that may already be an exclusive lock that would fail one of the concurrent transactions. The protection of an explicit lock like this should make sure that e.g. version 2 is not deleted while version 3 is being added, creating an inconsistency.

Reply:
First answer from postgres is that in the basic version, without explicit locks, the second transaction stalls behind the first one that gets to execute an `UPDATE`; it is not rolled back however but allowed to go ahead after the first commit.
Will try the add vs. remove concurrency use case now.

Reply:
Analysis of `PESSIMISTIC_READ` leads me to think it can cause deadlocks if two transactions both hold a read lock and try to update the value:
```
postgres=# UPDATE item SET version = 3 WHERE id='naledi';                            
ERROR:  deadlock detected
DETAIL:  Process 98 waits for ExclusiveLock on tuple (0,8) of relation 16384 of database 12994; blocked by process 83.
Process 83 waits for ShareLock on transaction 569; blocked by process 98.
HINT:  See server log for query details.
```
The deadlock is detected (by Postgres in this case) and aborts one of the two, so it may be the best solution here. It can also be escalated immediately to `PESSIMISTIC_WRITE` since we intend to update it, with lower concurrency as it queues the new `SELECT` requesting the same lock (but not the ones used for the read side of the API).

Reply:
We'll end up having locks at the API level too (think there should only ever be 1 processing of an item), so sounds like this is ok. Thanks for the detailed look.